source/blog/2022-11-04-annotations-to-attributes.md

```diff
@@ -0,0 +1,167 @@
+---
+title: "From annotations to attributes"
+authorName: Grégoire Paris
+authorEmail: postmaster@greg0ire.fr
+permalink: /2022/11/04/annotations-to-attributes.md
+---
+
+Last month, we migrated the tests of the ORM from annotations to attributes.
+Let us look back on what lead to this moment.
+
+## Annotations
+
+Let's go 20 years back in time. In 2002, Alex Buckley, a Specification lead for
+the Java language publishes [JSR-175][jsr-175], proposed to add user-defined
+annotations to the language, allowing to tag language elements with extra
+information. 2 years later, it gets approved and Java 1.5, also known as Java 5
+is released, [with support for annotations][annotation-guide].
+
+2 more years elapse and in 2004, Jano Suchal publishes [Addendum][addendum], a
+PHP library that adds support for "Docblock/JavaDoc" annotations, meaning that
+contrary to what is done in Java, Addendum annotations are contained inside
+phpdoc comments, like this:
+
+```php
+/** @test */
+function test_it_throws_on_invalid_argument(): void
+{}
+```
+
+That is because they are implemented in userland, without requiring a change in
+PHP itself.
+
+Doctrine ORM 2.0 is not released yet at that point, but [the library is used to
+build an annotation driver][addendum-common] in Doctrine 2 in early 2009.
+At that time, Doctrine was a project in a single repository, with `Common`,
+`DBAL` and `ORM` as top-level namespaces.
+[Addendum is replaced 6 months later][annotation-driver], with a new namespace
+under `Common` called `Annotations`.
+
+In the summer of 2010, Guilherme Blanco and Pierrick Charron submit
+[an RFC to add annotations support to PHP][annotations-rfc], but it
+[gets declined][mailing-list-annotations-rfc]. The RFC already mentions the
+need for annotations in PHPUnit, Symfony, Zend Framework, FLOW3 and of course,
+Doctrine.
+
+Late 2010, Doctrine 2 is tagged, and the single repository is split into 3
+repositories.
+
+Finally, in 2013, the namespace above is isolated in its own repository, and
+`doctrine/annotations` 1.0.0 is tagged.
+
+Today, the package is widely used in the PHP ecosystem and has a little short
+of 300 million downloads on Packagist, and is depended on by over 2 thousand
+packages, including major frameworks and tools. It is fair to say annotations
+have proven valuable to many users.
+
+[jsr-175]: https://www.cs.ubc.ca/~gregor/teaching/cpsc411/metadata-public-draft.html
+[annotation-guide]: https://docs.oracle.com/javase/1.5.0/docs/guide/language/annotations.html
+[addendum]: https://github.com/jsuchal/addendum
+[addendum-common]: https://github.com/doctrine/orm/commit/bcf0110249978c8fc705916d5a4f94216bb98b07
+[annotation-driver]: https://github.com/doctrine/orm/commit/9075f10bf5da3058f6ef82ff08e4783ff70424a4
+[annotations-rfc]: https://wiki.php.net/rfc/annotations
+[mailing-list-annotations-rfc]: https://externals.io/message/49733
+
+## Attributes
+
+The RFC mentioned above is only one among [many][rfc]. As mentioned before,
```

Comment:
the link reference should be `rfc-list`, not `rfc` (or it should be renamed below)